Kibana UI scaling guide (#2967)
elastic/docs-content: deploy-manage/production-guidance/kibana-traffic-scaling-considerations.md

1. Concurrency: how many users do you expect to be interacting with {{kib}} simultaneously, which is largely **CPU-bound**
2. Request and response size: how large (usually measured in bytes) are the requests and responses you expect {{kib}} to service, which is largely **memory-bound**

::::{important}
The nature of traffic is not only diverse, but also unpredictable. Traffic to {{kib}} often comes in short bursts or spikes that can overwhelm an underprovisioned {{kib}}. In production environments, an overwhelmed {{kib}} will typically manifest as 502 or 503 error responses from {{kib}}.

Load balancing helps to mitigate this bursty nature of traffic by horizontally scaling your {{kib}} deployments and improving {{kib}}'s availability. See the guide on [load balancing traffic](./kibana-load-balance-traffic.md).
::::

::::{important}
CPU and memory boundedness often interact in important ways. If CPU-bound activity is reaching its limit, memory pressure will likely increase as {{kib}} has less time for activities like garbage collection. If memory-bound activity is reaching its limit, there may be more CPU work to free claimed memory, increasing CPU pressure.

> ::::

Author: > what is someone supposed to take away from this?

My intention was just to call out that when one resource is under pressure another might also be. So I guess the take-away is something like:

Observe CPU and memory over time (as with the more sophisticated guide) to get an idea of where your bottleneck really is.

Consider these examples:

* {{kib}} to should serve 50 concurrent users: **10.5 vCPU** and **10GB** of memory which on {{ech}} and {{ece}} translates to: **2 {{kib}} instances of 8.4 vCPU and 8GB memory each** or **3 {{kib}} instances of 8.4 vCPU and 4GB memory each**
* {{kib}} to serve 100 concurrent users you would need **25.2 vCPU** and **20GB** of memory which on {{ech}} and {{ece}} translates to: **3 {{kib}} instances of 8.4 vCPU and 8GB memory each**.

::::{important}
This advice does not apply to scaling {{kib}} for task manager. If you intend to use {{kib}} alerting capabilities see [task manager scaling guidance](./kibana-task-manager-scaling-considerations.md).

> ::::

Reviewer: ++ I'd put this much earlier in the document so people know to redirect if they are primarily interested in task manager.

> **{{ece}}, {{ech}}, and {{eck}}** users can adjust {{kib}}'s memory by viewing their deployment and editing the {{kib}} instance's resource configuration. Note: size increments are predetermined.
> 
> **Self-managed** users must provision memory to the host that {{kib}} is running on as well as configure allocated heap. See [the guidance on configuring {{kib}} memory](./kibana-configure-memory.md). **Note:** Node.js suggests allocating 80% of available host memory to heap, assuming that Kibana is the only server process running on the (virtual) host. This allows for memory resources to be used for other activities, for example: allowing for HTTP sockets to be allocated.

Reviewer 2: as far as I know, to resize your deployment in eck, you can't do it this way because there's no UI. you have to do something totally different. couple of resources (I don't know details on how to do this personally):

https://www.elastic.co/docs/deploy-manage/deploy/cloud-on-k8s/manage-compute-resources
https://www.elastic.co/docs/deploy-manage/autoscaling/autoscaling-in-eck#k8s-stateless-autoscaling

suggest also adding links. added some (looks like the ech one has a bit of a gap for kibana unfortunately)

```suggestion
### Adjust resource allocations for {{kib}}

The way that you alter the resources allocated to your {{kib}} instance depends on your deployment type.

* **[{{ech}}](/deploy-manage/deploy/elastic-cloud/ec-customize-deployment-components.md) and [{{ece}}](/deploy-manage/deploy/elastic-cloud/configure.md):** Users can adjust {{kib}}'s memory by viewing their deployment and editing the {{kib}} instance's resource configuration. In these environments, size increments are predetermined.

* **{{eck}}:** (instruction)

* **Self-managed:** Users must provision memory to the host that {{kib}} is running on as well as configure allocated heap. See [the guidance on configuring {{kib}} memory](./kibana-configure-memory.md). 

	:::{note}
	Node.js suggests allocating 80% of available host memory to heap, assuming that {{kib}} is the only server process running on the (virtual) host. This allows for memory resources to be used for other activities, for example, allowing for HTTP sockets to be allocated.
	:::
```

**Serverless** manages {{kib}}'s resources automatically.

### A more sophisticated sizing strategy

::::{important}
On [serverless](../deploy/elastic-cloud/serverless.md) scaling and configuring {{kib}} is fully managed for you.
::::

Building on the simple strategy outlined above, we can make more precise adjustments to resource allocations. **Self-managed** users manage their CPU and memory allocations independently and can employ the strategy below to further tailor resources based on performance metrics.

#### Monitoring [_monitoring-kibana-metrics]

> In order to understand the impact of your usage patterns on **a single {{kib}} instance** use the Stack Monitoring feature. See [the guide for {{kib}} deployed on {{ech}} or {{ece}}](../monitor/stack-monitoring/ece-ech-stack-monitoring.md) or the [the guide for self-managed {{kib}}](../monitor/stack-monitoring/kibana-monitoring-self-managed.md).

Reviewer 2: this also works for eck. would suggest just linking to the top level for stack monitoring.

also: could people use autoops for this? should they?

```suggestion
In order to understand the impact of your usage patterns on a single {{kib}} instance, use the [stack monitoring](/docs/deploy-manage/monitor/stack-monitoring) feature.
```

> The rest of this guide will assume you have visibility into the following important metrics for a {{kib}} instance:
> 
> 1. Event loop delay (ELD) in milliseconds - this is a Node.js concept that roughly translates to the number of milliseconds by which processing of events is delayed due to CPU-intensive activities
> 2. Heap size in bytes - the amount of bytes currently held in memory dedicated to {{kib}}'s heap space
> 3. HTTP connections - the number of sockets that the Kibana server has open

Reviewer 2: ```suggestion
Using stack monitoring, you can gather the following metrics for your {{kib}} instance:

* **Event loop delay (ELD) in milliseconds:** A Node.js concept that roughly translates to the number of milliseconds by which processing of events is delayed due to CPU-intensive activities.
* **Heap size in bytes:** The amount of bytes currently held in memory dedicated to {{kib}}'s heap space.
3. **HTTP connections: ** The number of sockets that the {{kib}} server has open.
```

##### CPU [kibana-traffic-load-cpu-sizing]

> Event loop delay (ELD) is an important metric for understanding whether Kibana is engaged in CPU-bound activity.

Reviewer 2: ```suggestion
Event loop delay (ELD) is an important metric for understanding whether {{kib}} is engaged in CPU-bound activity.
```

> **As a general target, ELD should be at below ~200ms 95% of the time**. Higher delays may mean {{kib}} is CPU-starved. Sporadic increases above 200ms may mean that Kibana is periodically processing CPU-intensive activities like large responses from Elasticsearch, whereas consistently high ELD may mean Kibana is struggling to service tasks and requests.

Reviewer 2: ```suggestion
**As a general target, ELD should be at below ~200ms 95% of the time**. Higher delays may mean {{kib}} is CPU-starved. Sporadic increases above 200ms may mean that {{kib}} is periodically processing CPU-intensive activities like large responses from {{es}}, whereas consistently high ELD may mean {{kib}} is struggling to service tasks and requests.
```

Before increasing CPU resources, consider the impact of ELD on user experience. If users are able to use {{kib}} without the frustration that comes from a blocked CPU, provisioning additional CPU resources will not be impactful, although having spare resources in case of unexpected spikes is useful.

Monitoring {{kib}}'s ELD over time is a solid strategy for knowing when additional CPU resource is needed based on your usage patterns.

##### Memory [kibana-traffic-load-memory-sizing]

Heap size is an important metric to track. If {{kib}}'s heap size grows beyond the heap limit, {{kib}} will crash. By monitoring heap size, you can help ensure that {{kib}} has enough memory available.

> **Self-managed** users must provision memory to the host that {{kib}} is running on as well as configure allocated heap. See [the guidance on configuring {{kib}} memory](./kibana-configure-memory.md). **Note:** Node.js suggests allocating 80% of available memory to heap. This allows for memory resources to be used for other activities, for example: allowing for HTTP sockets to be allocated.

Reviewer 2: do we need this note twice? 
```suggestion
Self-managed users must provision memory to the host that {{kib}} is running on as well as configure allocated heap. See [the guidance on configuring {{kib}} memory](./kibana-configure-memory.md).

:::{note}
Node.js suggests allocating 80% of available memory to heap. This allows for memory resources to be used for other activities, for example: allowing for HTTP sockets to be allocated.
:::
```

## Basic scaling using number of concurrent users

Follow this strategy if you know the maximum number of expected concurrent users.

Start {{kib}} on **1 vCPU** and **2GB** of memory. This should comfortably serve a set of 10 concurrent users performing analytics activities like browsing dashboards.

If you are experiencing performance issues, you can scale {{kib}} vertically by adding the following resources for every 10 additional concurrent users:
* 1 vCPU
* 2GB of memory

These amounts are a safe minimum to ensure that {{kib}} is not resource-starved for common analytics use cases.

It is recommended to scale vertically to a maximum of **8.4 vCPU** and **8GB** of memory.

You should also combine vertical scaling with horizontal scaling to handle greater concurrency or bursty traffic. Refer to [](./kibana-load-balance-traffic.md) for guidance.

::::{important}
This advice does not apply to scaling {{kib}} for task manager. If you intend to use {{kib}} alerting capabilities, see [](./kibana-task-manager-scaling-considerations.md).
::::

### Scaling examples

> | Concurrent users | Total vCPU | Total memory | ECH and ECE deployment examples |

Reviewer 3: ```suggestion
| Concurrent users | Minimum vCPU | Minimum memory | ECH and ECE deployment examples |
```

Not sure about this one. I understand that "Total" is used here to refer to the total memory across instances that is required. But thought "Total" might be slightly confusing if the example numbers don't add up (2x8 = 16 != 10) and maybe minimum is slightly less confusing.

| --- | --- | --- | --- |

> | 50 | 5 vCPU | 10GB | • {{kib}} instance of 8.4 vCPU and 16GB memory in 1 availability zone (creates 2 8GB nodes per availability zone)<br><br>• {{kib}} instance of 8.4 vCPU and 8GB memory each across 2 availability zones<br><br>• {{kib}} instances of 8.4 vCPU and 4GB memory each across 3 availability zones |

Reviewer 3: ```suggestion
| 50 | 5 vCPU | 10GB | • {{kib}} size per zone of 16GB RAM and 8 vCPU in 1 availability zone (creates 2 x 8GB nodes)<br><br>• {{kib}} size per zone of 8GB RAM and up to 8 vCPU across 2 availability zones<br><br>• {{kib}} size per zone of 4GB RAM and up to 8 vCPU across 3 availability zones |
```

> | 100 | 10 vCPU | 20GB | • {{kib}} instance of 12 vCPU and 24GB memory in 1 availability zone (creates 3 8GB nodes per availability zone)<br><br>• {{kib}} instance of 8.4 vCPU and 8GB memory each across 3 availability zones<br><br>|

Reviewer 3: ```suggestion
| 100 | 10 vCPU | 20GB | • {{kib}} size per zone of 24 GB RAM and 12 vCPU in 1 availability zone (creates 3 x 8GB nodes)<br><br>• {{kib}} size per zone of 8GB RAM and up to 8 vCPU across 3 availability zones<br><br>|
```

Nits to bring this slightly closer to the copy on the UI, e.g. listing memory first and calling it "RAM", using 8vCPU instead of 8.4, using "size per zone" instead of "instance".

:::

## Advanced scaling using stack monitoring metrics

Building on the simple strategy outlined above, we can identify where {{kib}} is resource constrained more precisely. **Self-managed** and **{{eck}}** users manage CPU and memory allocations independently and can further tailor resources based on performance metrics.

### Gather usage information [_monitoring-kibana-metrics]

In order to understand the impact of your usage patterns on a single {{kib}} instance, use the [stack monitoring](../monitor/stack-monitoring.md) feature.

Using stack monitoring, you can gather the following metrics for your {{kib}} instance:

* **Event loop delay (ELD) in milliseconds:** A Node.js concept that roughly translates to the number of milliseconds by which processing of events is delayed due to CPU-intensive activities.
* **Heap size in bytes:** The amount of bytes currently held in memory dedicated to {{kib}}'s heap space.
* **HTTP connections:** The number of sockets that the {{kib}} server has open.

> #### Scale CPU using ELD metrics [kibana-traffic-load-cpu-sizing]

Reviewer 2: ```suggestion
### Scale CPU using ELD metrics [kibana-traffic-load-cpu-sizing]
```

Event loop delay (ELD) is an important metric for understanding whether {{kib}} is engaged in CPU-bound activity.

**As a general target, ELD should be below ~220ms 95% of the time**. Higher delays may mean {{kib}} is CPU-starved. Sporadic increases above 200ms may mean that {{kib}} is periodically processing CPU-intensive activities like large responses from {{es}}, whereas consistently high ELD may mean {{kib}} is struggling to service tasks and requests.

Before increasing CPU resources, consider the impact of ELD on user experience. If users are able to use {{kib}} without the frustration that comes from a blocked CPU, provisioning additional CPU resources will not be impactful, although having spare resources in case of unexpected spikes is useful.

Monitoring {{kib}}'s ELD over time is a solid strategy for knowing when additional CPU resource is needed based on your usage patterns.

> #### Scale memory using heap size metrics [kibana-traffic-load-memory-sizing]

Reviewer 2: ```suggestion
### Scale memory using heap size metrics [kibana-traffic-load-memory-sizing]
```
